[Multi Cluster] Federation and Cluster Management intro (#174)
mspnp/aks-secure-baseline: README.md

# Azure Kubernetes Service (AKS) Multi Cluster

> This reference implementation will go over some design decisions from the baseline to detail them as a well as incorporate some new _recommended infrastructure options_ for a Multi Cluster architecture. In this oportunity, this implementation and document are meant to guide the multiple distinct teams as introduced in the [AKS Baseline](https://github.com/mspnp/aks-secure-baseline) through the process of expanding from a single cluster to multiple clusters solution with a foundamental driver in mind which is **High Availability**.

Reviewer: ```suggestion
This reference implementation will go over some design decisions from the baseline to detail them as a well as incorporate some new _recommended infrastructure options_ for a multi-cluster AKS architecture. This implementation and document are meant to guide the multiple distinct teams as introduced in the [AKS Baseline](https://github.com/mspnp/aks-secure-baseline) through the process of expanding from a single cluster to a multi-cluster solution with a fundamental objective of **High Availability**.
```

Author: I agree 100% that some content in here is too much for an overview, and we should move that to other sections. We plan to keep moving things around in the following days. 

Thanks for the great co-authoring edition that improved how it reads 10X 😹

> Throughout the reference implementation, you will see reference to _Contoso Bicycle_. They are a fictional small and fast-growing startup that provides online web services to its clientele on the east coast US-based. This narrative provides grounding for some implementation details, naming conventions, etc. You should adapt as you see fit.

Reviewer: ```suggestion
Throughout the reference implementation, you will see reference to _Contoso Bicycle_. They are a fictional, small, and fast-growing startup that provides online web services to its clientele on the east coast of the United States. This narrative provides grounding for some implementation details, naming conventions, etc. You should adapt as you see fit.
```

### Intro

> The app team works on a cluster strategically located in the `East US 2` region as this is where most of their customer base can be found. They have been operating this solo AKS cluster for a quite some time [following the recommended baseline practices to work with a single AKS cluster, and they learned them all by heart](https://github.com/mspnp/aks-secure-baseline).

Reviewer: ```suggestion
The app team works on a cluster strategically located in the `East US 2` region as this is where most of their customer base can be found. They have been operating this single AKS cluster for a quite some time [following Microsoft's recommended baseline architecture](https://github.com/mspnp/aks-secure-baseline).
```

> Thanks to baseline the cluster is already available from different _Zones_ within the same region, now they realize that provided `East US 2` went fully down, they are out of game. Even though the SLA(s) are acceptable for their business continuity plan, they are starting to think what are the options, and how their stateless application (Application ID: a0042) could increase its availability in case of an incident of the magnitude mentioned above. They started conversations with the business unit (BU0001) to increment the number of clusters by one. In other words, they are proposing to go over a Multi Cluster infrastructure solution in which multiple intances of the same application could live.

Reviewer: ```suggestion
The baseline the cluster was already available from different _Zones_ within the same region, but now they realize that if `East US 2` went fully down, zone coverage is not sufficient. Even though the SLA(s) are acceptable for their business continuity plan, they are starting to think what their options are, and how their stateless application (Application ID: a0042) could increase its availability in case of a complete regional outage. They started conversations with the business unit (BU0001) to increment the number of clusters by one. In other words, they are proposing to move to a multi-cluster infrastructure solution in which multiple instances of the same application could live.
```

Author: <3

> This architectural decision will have multiple implications for the Contoso Bicycle organization. It is not just about repeating everything they have done before but planning how efficiently share Azure resources as well as detect those that need to be added, how they are going to deploy more than one cluster as well as operate them, decide to which specific regions they deploy, and many more considerations striving for High Availability.

Reviewer: ```suggestion
This architectural decision will have multiple implications for the Contoso Bicycle organization. It is not just about repeating everything they have done before but instead planning how to efficiently share Azure resources as well as detect those that need to be added; how they are going to deploy more than one cluster as well as operate them; decide to which specific regions they deploy; and many more considerations striving for high availability.
```

### Federation

> The business unit (BU0001) approves the creation of a second cluster that could help balacing the traffic but mainly to serve as a hot backup, they are a bit worried on the required engineering effort though. The same application (Application Id: a0042) is about to span into multiple clusters, so there is a desire to find out a good mechanism for its configuration management. With that in mind, the app team is looking at what _Federation_ approaches they could follow to run different intances of the exact same app in different clusters.

Reviewer: ```suggestion
The business unit (BU0001) approves the creation of a second cluster that could help balance the traffic but mainly to serve as a hot backup; they are a bit worried about the required engineering effort though. The same application (Application Id: a0042) is about to span into multiple clusters, so there is a desire to find a good mechanism for its configuration management. With that in mind, the app team is looking at what _federation_ approaches they could follow to run different instances of the exact same app in different clusters.
```

> They know that at this point things must be kept simply, in fact they could run these two application instances from these two clusters with just a bunch of useful scripts. But they want to be sure that the selected approach is not going to be adding impediments that could prevent from scaling out their fleet of clusters down to road if there was a requirment to scale up the amount application instances.

Reviewer: ```suggestion
They know that at this point things must be kept simple; in fact they could run these two application instances from these two clusters with just a bunch of useful scripts. But they want to be sure that the selected approach is not going to be adding impediments that could prevent from scaling out their fleet of clusters down to road if there was a requirement to scale out even further.
```

> Depending on how federation is implemetned it could open a door in which a single command execution has an instant ripple effect into all your clusters. While running clusters separately like silos could keep you safe from it, the cost could be really hight by struggling at scaling the number of clusters in the future.

Reviewer: ```suggestion
Depending on how federation is implemented it could open a door in which a single command execution has an instant ripple effect into all your clusters. While running clusters separately like silos could keep you safe from the same, but the cost could be really high to scale the number of clusters in the future.
```

> They know that there is specialized tooling out there that helps managing a centralized control plane to push the workload(s) behavior top to bottom reacting to special events like _region down_ but they want to proceed with caution in this area for now.

Reviewer: ```suggestion
They know that there is specialized tooling out there that helps manage a centralized control plane to push the workload(s) behavior top to bottom reacting to special events like a regional outage but they want to proceed with caution in this area for now.
```

> Given that this Reference Implementation provides with a middle ground solution in which an organization could build the basis for the future without this being a weight on their shoulders for just two clusters. Therefore, the recommendation is to manage the workload manifests sperately per instance from a central _Federation_ git repository in combination with a CI/CD pipeline. The latter is not implemenated as part of this reference implementation.

Reviewer: ```suggestion
Given that this reference implementation provides a middle ground solution in which an organization could build the basis for the future without this being a weight on their shoulders for just two clusters. Therefore, the recommendation is to manage the workload manifests separately per instance from a central _federation_ git repository in combination with a CI/CD pipeline. The latter is not implemented as part of this reference implementation.
```

### Multi cluster management in multiple regions

> The new selected location is `Central US` that happens to be the Azure paired region for `East US 2`. Now the networking team in conjunction with the app team are closely working together to understand what is the best way for laying down the new fleet of clusters.

Reviewer: ```suggestion
The new selected location is `Central US` which is the Azure paired region for `East US 2`. Now the networking team in conjunction with the app team are closely working together to understand what is the best way for laying down the new cluster.
```

> All in all, the team resolution is to have CI/CD pipelines, traffic management and centralized gitops as well as centralize the git repos containing the workloads manifest, and a single declarative stamp for the cluster creation with different param files per region.

Reviewer: ```suggestion
All in all, the team resolution is to have CI/CD pipelines, traffic management, and centralized GitOps as well as centralize the git repos containing the workload manifests, and a single declarative stamp for the cluster creation with different parameter files per region.
```

![The federation diagram depicting the proposed cluster fleet topology running different instances of the same application from them.](images/aks-cluster-mgmnt-n-federation.png)

> :bulb: Multi Cluster and Federation's repos could be a monorepo or multiple repos as displayed from the digram above. In this reference implementation, the workload manifests, and ARM templates are shipped together from a single repo.

## Azure Architecture Center guidance

This project has a companion set of articles that describe challenges, design patterns, and best practices for a secure AKS cluster. You can find this article on the Azure Architecture Center at [Azure Kubernetes Service (AKS) Baseline Cluster](https://aka.ms/architecture/aks-multicluster). If you haven't reviewed it, we suggest you read it as it will give added context to the considerations applied in this implementation. Ultimately, this is the direct implementation of that specific architectural guidance.

## Architecture

> **This architecture is infrastructure focused again**, more so than on workload. It concentrates on two AKS clusters, including concerns like multi region deployments, the desired state of the cluster, geo replication, network topologies and more.

Reviewer: ```suggestion
**This architecture is infrastructure focused**, more so than on workload. It concentrates on two AKS clusters, including concerns like multi-region deployments, the desired state of the clusters, geo-replication, network topologies, and more.
```

> The implementation presented here like in the baseline is the _minimum recommended starting (baseline) for a multiple AKS cluster solution_. This implementation integrates with Azure services that will deliver geo replication, a centralized observability appraoch, a network topology that is going go with multi-regional growth for real, and balancing the all traffic as well.

Reviewer: ```suggestion
The implementation presented here, like in the baseline, is the _minimum recommended starting (baseline) for a multiple AKS cluster solution_. This implementation integrates with Azure services that will deliver geo-replication, a centralized observability approach, a network topology that is going go with multi-regional growth, and an added benefit of additional traffic balancing as well.
```

> The material here will try not to be dense or at least focused exclusevely on the multi regional growth. We strongly encourage you to allocate time to walk through the AKS Basline and later all the instructions provided in here, with a mind to learning. We do NOT provide any "one click" deployment here. However, once you've understood the components involved and identified the shared responsibilities between your team and your great organization, it is encouraged that you build suitable, auditable deployment processes around your final infrastructure.

Reviewer: ```suggestion
The material here will try not to be dense or at least focused exclusively on the multi-regional growth. We strongly encourage you to allocate time to walk through the AKS Baseline and later all the instructions provided in here, with a mind to learning. We do NOT provide any "one click" deployment here. However, once you've understood the components involved and identified the shared responsibilities between your team and your great organization, it is encouraged that you build suitable, auditable deployment processes around your final infrastructure.
```

> The material here will try not to be dense or at least focused exclusevely on the multi regional growth. We strongly encourage you to allocate time to walk through the AKS Basline and later all the instructions provided in here, with a mind to learning. We do NOT provide any "one click" deployment here. However, once you've understood the components involved and identified the shared responsibilities between your team and your great organization, it is encouraged that you build suitable, auditable deployment processes around your final infrastructure.

Reviewer: > The material here will try not to be dense or at least focused exclusively on the multi-regional growth

Not sure what that sentence is really trying to say.

Author: tried to clarify a bit the intent from [c36a4](https://github.com/mspnp/aks-secure-baseline/pull/174/commits/c36a4695126e483a607c8684d2c9f9f1d6b621e5)
